transaction/std/src/transaction_builder.rs

```diff
@@ -139,7 +142,72 @@ impl<FPR: FogPubkeyResolver> TransactionBuilder<FPR> {
     /// * `input_credentials` - Credentials required to construct a ring
     ///   signature for an input.
     pub fn add_input(&mut self, input_credentials: InputCredentials) {
-        self.input_credentials.push(input_credentials);
+        self.input_materials
+            .push(InputMaterials::Signable(input_credentials));
+    }
+
+    /// Add a pre-signed Input to the transaction, also fulfilling any
+    /// requirements imposed by the signed rules, so that our transaction
+    /// will be valid.
+    ///
+    /// Note: Before adding a signed_contingent_input, you probably want to:
+    /// * validate it (call .validate())
+    /// * check if key image appreared already (call .key_image())
+    /// * provide merkle proofs of membership for each ring member (see
+    ///   .tx_out_global_indices)
+    ///
+    /// # Arguments
+    /// * `signed_contingent_input` - The pre-signed input we are adding
+    pub fn add_presigned_input(
```

Comment:
Tying into my comment above re: creating an InputRule trait for reach individual rule- it could perhaps be used here as well. This would make adding a new input rule easier for another engineer. Right now, you'd have to know to add to the InputRules verify method then add the verification logic here as well.

Reply:
I think it's okay for engineers to have to know that if they add a new feature, they have to adjust multiple places in the code. Right now if you add a new transaction validation rule, you have to modify the validation module, and also the builder to ensure that the builder makes things that satisfy the rule. Additionally, it's possible that not all the rules can be enforced in isolation, they may tie into each other somehow, and then a trait probably wouldn't work.

Also, it's not clear if this API is actually going to work in general, it has been proposed that we should have rules that are like "or" rules that can be satisfied in several possible ways, potentially numerous. Then the builder will not be able to infer which way you have in mind, so you would have to use `add_presigned_input_raw` for that, and then some kind of `add_output_raw` function as well (which doesn't exist yet). Or we would need some other more helpful API here.
So given that we might be headed in that direction I'm not sure if a trait is going to work out.